### ISSUE TITLE
create a video file encoded using WebCodecs 

### ISSUE THREAD

**Issue by tlecoz** [gpac/mp4box.js#243]
Hello ! 

I'm trying to generate a video file using mp4box & webcodec directly inside the browser, without nodejs & ffmpeg. It almost works but my video-file contains only 2-3 frames instead of 300 and I don't understand at all what's happening. 

The problem come from the mp4box part because the same code works well if I use ffmpeg (with nodejs) as a muxer instead of mp4box. 

I'm not an expert at all with mp4box, muxers and stuffs like that. Few months ago I  didn't even know the word "muxer"  ( 😅 ) so please do not assume that I understand what is behind the hood. I'm just a web-developer , not a video-master ;)  

Webcodec is not released yet. It will be released next month but you can use it now if you go to chrome://flags , then search "Experimental Web Platform features" and enable it. 

More infos about webcodecs here : https://github.com/w3c/webcodecs

Here is my experimentation with mp4box (it's a minimal example, it couldn't be smaller) : 

```
  <html>
    <script src="mp4box.all.min.js" ></script>
    <body>
        <script>
			
            let w = 1920;
            let h = 1080;
            let durationInMillisecond = 5000;
            let fps = 60;
            let frameTimeInMillisecond = 1000 / fps; 
	    let totalFrames = Math.floor(durationInMillisecond / frameTimeInMillisecond) ;
            let chunkCount = 0;
            let videoEncoder;
            let encoderClosed = false;


            //------ Mp4box setup --------

            let oneSecondInMillisecond = 1000;
            let timescale = 1000;

            let trackOptions = {
                timescale:(oneSecondInMillisecond * timescale),
                duration:(durationInMillisecond * timescale),
                width:w,
                height:h,
                nb_samples:totalFrames,
            }

            let sampleOptions = {
                duration:(frameTimeInMillisecond * timescale)
            } 

            let file = MP4Box.createFile();
            let track = file.addTrack(trackOptions);
            
            //--- WebCodec videoEncoding -----
	   async function buildAndConfigureEncoders() {
			    let videoEncoder = new VideoEncoder({
                                  output: (encodedChunk,config)=>{ 
                        
                                       file.addSample(track,new Uint8Array(encodedChunk.data),sampleOptions);
                        
                                       chunkCount++;
                                       if (chunkCount==totalFrames) {
                                          videoEncoder.close();
                                          encoderClosed = true;
                                          file.save("test.mp4");
                                         console.log("completed !");
                                       }
                        
                                },
                               error: (error)=>{
                                     console.log("onCodecError ",error);
                                }
                         });
			  
                
                         await videoEncoder.configure({
                            codec : 'avc1.42001E',
                            width:w,
                            height:h,
                            hardwareAcceleration:"deny",
                            avc:{format:"annexb"} // format can be "annexb" or "avc" but mp4box doesn't seem to support "avc"
                        });
				
			 
		  return videoEncoder;
	    }	
            
            
            //--- create a simple movie inside a canvas and send frames to webcodec.VideoEncoder ----
            let canvas = document.createElement("canvas");//new OffscreenCanvas(w, h);
            canvas.width = w;
            canvas.height = h;
            document.body.appendChild(canvas)
            let ctx = canvas.getContext("2d");

            function drawFrame(progress){
                if(encoderClosed) return;
                
                ctx.fillStyle = "#ffffff";
                ctx.fillRect(0,0,w,h);
                ctx.fillStyle = "#000000";
                ctx.fillRect(Math.round((w-50)*progress),200,50,50);
                
                createImageBitmap(canvas).then((bmp)=>{
                    let videoFrame = new VideoFrame(bmp,{timestamp:durationInMillisecond*progress});
                    videoEncoder.encode(videoFrame);
                    videoFrame.close();
                });			
            }

            async function start() {
		 videoEncoder = await buildAndConfigureEncoders();
                
	 	 let i = 0;
                  while (i++<totalFrames) drawFrame(i/totalFrames)		
	   }
			
	   start();

        </script>
    </body>
</html>
```
  
 You can see the result here : 
  https://beginfill.com/webcodec_mp4box/test.mp4 
  
  (that's weird, online it doesn't move at all, but if I use the video-player called VLC I can see the first 2-3 frames) 
  
  
  Can someone please help me to get it working as exected ? 
  
  Thank you for reading
  
[test_webcodec_videoEncoder_and_mp4box_muxer.zip](https://github.com/gpac/mp4box.js/files/7067512/test_webcodec_videoEncoder_and_mp4box_muxer.zip)

![image](https://user-images.githubusercontent.com/272817/131141210-26643dd4-1abf-4ea8-bc83-87bb80c725a8.png)

**Comment by cconcolato** [comment]
Here is a working version:
```<html>
    <script src="../dist/mp4box.all.js" ></script>
    <body>
        <script>
			
			
	
            let w = 1920;
            let h = 1080;
            let durationInMillisecond = 1000;
            let fps = 60;
            let frameTimeInMillisecond = 1000 / fps; 
			let totalFrames = Math.floor(durationInMillisecond / frameTimeInMillisecond) ;
            let chunkCount = 0;
            let videoEncoder;
            let encoderClosed = false;

            //------ Mp4box setup --------
            let oneSecondInMillisecond = 1000;
            let timescale = 1000;

            let trackOptions = {
                timescale:(oneSecondInMillisecond * timescale),
                width:w,
                height:h,
                nb_samples:totalFrames,
            }

            let sampleOptions = {
                duration: (frameTimeInMillisecond * timescale)
            } 

            let file = MP4Box.createFile();
            let track = null;

            //--- WebCodec videoEncoding -----
			async function buildAndConfigureEncoders() {
			    let videoEncoder = new VideoEncoder({
                   output: (encodedChunk,config)=>{ 
                        let ab = new ArrayBuffer(encodedChunk.byteLength);
                        encodedChunk.copyTo(ab);
                        if (track === null) {
                            trackOptions.avcDecoderConfigRecord = config.decoderConfig.description;
                            track = file.addTrack(trackOptions);
                        }
                        sampleOptions.dts = encodedChunk.timestamp*1000;
                        sampleOptions.cts = encodedChunk.timestamp*1000;
                        sampleOptions.is_sync = (encodedChunk.type === "key");
                        console.log(encodedChunk.timestamp);
                        file.addSample(track,ab,sampleOptions);
                        
                        chunkCount++;
                        if (chunkCount == 50) {
                            videoEncoder.close();
                            encoderClosed = true;
                            file.save("test.mp4");
                            console.log("completed !");
                        }
                        
                    },
                    error: (error)=>{
                        console.log("onCodecError ",error);
                    }
                });
			  
                
                await videoEncoder.configure({
                    codec : 'avc1.42001E',
                    width:w,
                    height:h,
                    hardwareAcceleration:"prefer-hardware",
                    avc:{format:"avc"} 
                });
				
			 
			  return videoEncoder;
			}	
            


            
            //--- create a simple movie inside a canvas and send frames to webcodec.VideoEncoder ----
            let canvas = document.createElement("canvas");//new OffscreenCanvas(w, h);
            canvas.width = w;
            canvas.height = h;
            document.body.appendChild(canvas)
            let ctx = canvas.getContext("2d");

            function drawFrame(progress){
                if(encoderClosed) return;
                
                ctx.fillStyle = "#ffffff";
                ctx.fillRect(0,0,w,h);
                ctx.fillStyle = "#000000";
                ctx.fillRect(Math.round((w-50)*progress),200,50,50);
                
                createImageBitmap(canvas).then((bmp)=>{
                    let videoFrame = new VideoFrame(bmp,{timestamp:durationInMillisecond*progress});
                    videoEncoder.encode(videoFrame);
                    videoFrame.close();
                });			
            }

            async function start() {
				videoEncoder = await buildAndConfigureEncoders();
                
				let i = 0;
                while (i++<totalFrames) drawFrame(i/totalFrames)		
			}
			
			start();

            
            

			


        </script>
    </body>
</html>```

Please use the latest mp4box.js. Note that the code is not stable and still quite hacky. 
I also have a working version of https://w3c.github.io/webcodecs/samples/capture-to-file/capture-to-file.html recording to AVC/MP4. 

@dalecurtis it's not clear to me how to control the timescale of timestamps returned by the WebCodecs API. This example seems to be returning milliseconds while the capture example returns microseconds. Any documentation?

**Comment by dalecurtis** [comment]
Timestamps should always be in microseconds, there's no timescale control at present. @tguilbert-google to investigate if there's a bug in MediaStreamTrackProcessor where it's not vending the right scale.

Here are some helpful links:
* https://web.dev/webcodecs/ is the article that just went out.
* Some samples are here https://w3c.github.io/webcodecs/samples/
* wpt tests are here https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/external/wpt/webcodecs/
* spec: https://w3c.github.io/webcodecs/

Some notes on the above:

```
createImageBitmap(canvas).then((bmp)=>{
  let videoFrame = new VideoFrame(bmp,{timestamp:durationInMillisecond*progress});
  videoEncoder.encode(videoFrame);
  videoFrame.close();
});	
```

Can just be

```
let videoFrame = new VideoFrame(canvas);
videoEncoder.encode(videoFrame);
videoFrame.close();
```

**Comment by cconcolato** [comment]
Thank you. I also saw that the samples use mp4box.js. Glad it was useful.

**Comment by dalecurtis** [comment]
Thank you for the awesome tool :)

**Comment by tlecoz** [comment]
Thank you so much to everyone for all your big work ! You are wonderfull, all of you ! 
Thank you !

**Comment by tlecoz** [comment]
@cconcolato  Where can I find the file 

`<script src="../dist/mp4box.all.js" ></script>` 

I don't find it on github 

Thank you again

**Comment by cconcolato** [comment]
Either you download the repo and run `grunt`, or you get it from https://gpac.github.io/mp4box.js/dist/mp4box.all.js

**Comment by tlecoz** [comment]
Thank you !

**Comment by Djuffin** [comment]
@cconcolato all timestamps in webcodecs are supposed to be in microseconds.
Where are you getting a frame or a chunk with a timestamp in milliseconds? 
It would be a pretty serious bug that we'd like to fix ASAP.

**Comment by tlecoz** [comment]
@Djuffin I may be wrong but I think the problem come from my code. 
This part : 

```
createImageBitmap(canvas).then((bmp)=>{
                    let videoFrame = new VideoFrame(bmp,{timestamp:durationInMillisecond*progress});
                    videoEncoder.encode(videoFrame);
                    videoFrame.close();
                });
```		

Until now I always used timestamp in milliseconds when I create the VideoFrame object and it works well when I mux the encodedchunks  with ffmpeg.

**Comment by tlecoz** [comment]
Hello @cconcolato @dalecurtis @tguilbert-google @Djuffin  ! 

I tryed to create a complete example of what webcodec can do that involve VideoDecoder, VideoEncoder, AudioDecoder and AudioEncoder in the same project. 

I wanted to create a minimal example without any classes, directly in the html. 
Well I maybe failled a bit because it represent something like 700 lines of code 😅 but I tryed to make it as readable as possible. 
I know it could be refactorized , some parts are redundant but I wanted something easier to follow for those who never touched to webcodec / mp4Box before. 

Unfortunately, I have some problems. 
Some of them concerns Mp4Box, other concerns WebCodec itself. 

@Webcodec team : 

- videoEncoderConfig : if I set "hardwareAcceleration" property to "prefer-software" , if I try to read the mp4 created by mp4box it stop after 3 secondes
 
 BUT 
 
 if I set this property to "prefer-hardware", the mp4 created by MP4Box works as expected 
 So there is definitly a bug somewhere inside webcodec. 
 					   
- audioEncoderConfig : the "codec" property doesn't match with MP4Box.audiotrack.codec. There is no example at all on the internet using AudioEncoder, can you give me a valid codec please
     
@cconcolato : 
- the mp4 generated is readable but the "playhead" in the "seekbar" stay at 0 (the time is updated correctly but not the "playhead") 

- Can you show me how to handle audio with the video please ? 
   Maybe it's the exact same thing than for the video but I suppose it's not. 
   I'm unable to try without a valid codec string for Webcodec.AudioEncoder 
		 
   For now, I added a variable "handleAudioEncoding" defined to "false" at the beginning of the code in order to skip this step
   and launch the saveFile function at the end of video-encoding 
   
   
Here is my code : 

```
<html>
    <head>
        <script src="mp4box.all.js" ></script>
    </head>
    <body>
        <script>
            var url = "videoWithSound.mp4";
            
            var handleAudioEncoding = false;


            
            var muxStarted = false;

            //---- MP4BOX VARIABLES
            var nbSampleMax = 30; // nb samples extracted each time
            var nbSampleTotal = 0;
            var countSample = 0;
            var file = null;
            var waitingFrame = false;
            var stopped = true;
            var videoTrack = null;
            var audioTrack = null;
            var outputFile = null;
            var oneSecondInMicrosecond = 1000000;

            //---- VIDEO DECODING VARIABLES
            var videoDuration = 0;
            var videoDecoder = null;
            var videoFrames = [];
            var videoNbSample;
            var decodedVideoFrameCount = 0;
            var processingVideo = false;
            var videoFramerate;
            var videoW;
            var videoH;
            

            //---- AUDIO DECODING VARIABLES
            var audioDecoder = null;
            var audioSampleLength = 1024;
            var audioSamplerate;
            var audioChannelCount;
            var audioNbSample;
            var audioFrames = [];
            var decodedAudioFrameCount = 0;
            var leftChannelOptions = {planeIndex:0,frameOffset:0,frameCount:audioSampleLength};
            var rightChannelOptions = {planeIndex:1,frameOffset:0,frameCount:audioSampleLength};
            var processingAudio = false;

            //----- VIDEO ENCODER VARIABLES 
            var videoEncoder = null;
            var encodingFrameDistance = 5// difference between last reading frame and last encoding frame
                                          // |-> we need to slow down the process in order to avoid a saturation of VideoEncoder
                                          // |==> when VideoEncoder is saturated by the data we send, the process became much more slower
                                          // |===> so, I'll pause the video decoding/reading when VideoDecoder progression became too late 
            
            var waitingVideoReading = false;
            var encodedVideoFrameCount = 0;
            var encodingVideoTrack = null;
            var videoFrameDurationInMicrosecond;
            var encodingVideoScale = 0.5; //we will transcode the video in another size ()
            var outputW;
            var outputH;

            //------ AUDIO ENCODING VARIABLES 
            var audioEncoder = null;
            var waitingAudioReading = false;
            var encodingAudioTrack = null;
            var encodedAudioFrameCount = 0;


            var output = document.createElement("canvas");
            output.width = outputW;
            output.height = outputH;
            var ctx = output.getContext("2d");
            
            document.body.appendChild(output);



            //---- Utility functions  --
            
            let freeMemory = ()=>{
                if(processingVideo){
                    if(decodedVideoFrameCount>0){
                        //let trackId = videoTrack.id;
                        //let lastSampleUsed = decodedVideoFrameCount-1;
                        file.releaseUsedSamples(videoTrack.id,decodedVideoFrameCount-1);
                    }
                    return
                }
                if(processingAudio){
                    if(decodedAudioFrameCount>0){
                        file.releaseUsedSamples(audioTrack.id,decodedAudioFrameCount-1);
                    }
                    return
                }
                
            }

            let onVideoDemuxingComplete = ()=>{  
                console.log("video demux complete"); 
                videoDecoder.close();

                processingVideo = false;
                
                if(audioTrack && handleAudioEncoding){
                    
                    setupAudioEncoder({
                        codec: audioTrack.codec,
                        sampleRate:audioTrack.audio.sample_rate,
                        numberOfChannels:audioTrack.audio.channel_count,
                        bitrate:audioTrack.bitrate
                    })

                    setupAudioDecoder({
                        codec: audioTrack.codec,
                        sampleRate:audioTrack.audio.sample_rate,
                        numberOfChannels:audioTrack.audio.channel_count
                    })
                    
                    getNextSampleArray();
                }

            }

            let saveFile = ()=>{
                outputFile.save("test.mp4");
                console.log("file saved !");
            }

            let onAudioDemuxingComplete = ()=>{
                console.log("onAudioDemuxCompleted !")
                audioDecoder.close();
            }

            let onAudioEncodingComplete = ()=>{
                audioEncoder.close();
                saveFile();
            }

            let onVideoEncodingComplete = ()=>{
                console.log("video encoding complete !")
                videoEncoder.close() //<=== in chrome 93 I must add a timeout in order to close the videoEncoder without error (ok in canary) 

                if(!audioTrack || !handleAudioEncoding) saveFile();
            }



            let continueReading = ()=>{
                if(processingVideo){
                    waitingVideoReading = decodedVideoFrameCount - encodedVideoFrameCount > encodingFrameDistance;
                    
                    console.log("VIDEO decodedFrameCount:",decodedVideoFrameCount+" VS encodedFrameCount:"+encodedVideoFrameCount)
                    
                    if(waitingVideoReading == false) readNextFrame();
                    else console.log("waiting videoEncoder")

                    return;
                }

                if(processingAudio){
                    waitingAudioReading = decodedAudioFrameCount - encodedAudioFrameCount > encodingFrameDistance;

                    console.log("AUDIO decodedFrameCount:",decodedVideoFrameCount+" VS encodedFrameCount:"+encodedVideoFrameCount)
                    
                    if(waitingAudioReading == false) readNextFrame();
                    else console.log("waiting audioEncoder")
                }
            }
            
            let onVideoFrameReadyToUse = (imageBitmap)=>{ 

                createImageBitmap(imageBitmap,0,0,videoW,videoH,{resizeWidth:outputW,resizeHeight:outputH,resizeQuality:"high"}).then((bmp)=>{
                    ctx.drawImage(bmp,0,0);

                    let timestamp = videoFrameDurationInMicrosecond * decodedVideoFrameCount;
                    console.log(timestamp)

                    let videoFrame = new VideoFrame(bmp,{timestamp:timestamp});
                    videoEncoder.encode(videoFrame);
                    videoFrame.close();
                    bmp.close();


                    decodedVideoFrameCount++;

                    if(decodedVideoFrameCount == nbSampleTotal) onVideoDemuxingComplete();
                    else continueReading();    

                })

                
                imageBitmap.close();
                
                
                
                
            }

            let onAudioFrameReadyToUse = (audioFrame)=>{

                /*
                //just an example to expose how to get audio-sample-buffers from AudioData object :
                //
                //let leftChannel = new Float32Array(audioSampleLength * 4);
                //audioFrame.copyTo(leftChannel,leftChannelOptions);
                //
                //let rightChannel = leftChannel;
                //if(audioFrame.numberOfChannels > 1){
                //    rightChannel = new Float32Array(audioSampleLength * 4);
                //    audioFrame.copyTo(rightChannel,rightChannelOptions);
                //}

                
                */
               
                
                audioEncoder.encode(audioFrame);
                audioFrame.close();

                decodedAudioFrameCount++;

                if(decodedAudioFrameCount == nbSampleTotal) onAudioDemuxingComplete();
                else readNextFrame();
            }


            let getNextSampleArray = ()=>{
                if(!stopped || !muxStarted || (processingVideo && countSample == nbSampleTotal) ) return;

                stopped = false;
                file.start();
            }

            let readNextFrame = ()=>{
                
                if(processingVideo){

                    if(videoFrames.length == 0){
                        if(decodedVideoFrameCount > 0){
                            
                            if(!waitingFrame){
                                waitingFrame = true;
                                //console.log("call getNextSampleArray ",decodedVideoFrameCount)
                                getNextSampleArray();
                            }
                        }
                    }else{ 
                        
                        
                        onVideoFrameReadyToUse(videoFrames.shift());
                    }
                    return
                }

                //-----------

                if(processingAudio){
                    if(audioFrames.length == 0){
                        if(decodedAudioFrameCount > 0){
                            if(!waitingFrame){
                                waitingFrame = true;
                                //console.log("call getNextSampleArray ",decodedAudioFrameCount)
                                getNextSampleArray();
                            }
                        }
                    }else{ 
                       
                        onAudioFrameReadyToUse(audioFrames.shift());
                    }
                    return
                }
               

            }

            //---- setup videoEncoder -----

            outputFile = MP4Box.createFile();

           

            let setupVideoEncoder = (config)=>{

                let videoEncodingTrackOptions = {
                    timescale:oneSecondInMicrosecond,
                    w:outputW,
                    h:outputH,
                    nb_samples:videoNbSample,
                    avcDecoderConfigRecord:null
                }

                let videoEncodingSampleOptions = {
                    duration:videoFrameDurationInMicrosecond,
                    dts:0,
                    cts:0,
                    is_sync:false
                }

                videoEncoder = new window["VideoEncoder"]({
                    output:(encodedChunk,config)=>{

                        if(encodingVideoTrack == null){
                            videoEncodingTrackOptions.avcDecoderConfigRecord = config.decoderConfig.description;
                            encodingVideoTrack = outputFile.addTrack(videoEncodingTrackOptions);
                        }

                        let buffer = new ArrayBuffer(encodedChunk.byteLength);
                        encodedChunk.copyTo(buffer);

                        videoEncodingSampleOptions.dts = videoEncodingSampleOptions.cts = encodedChunk.timestamp;
                        videoEncodingSampleOptions.is_sync = encodedChunk.type == "key";

                        outputFile.addSample(encodingVideoTrack,buffer,videoEncodingSampleOptions);

                        encodedVideoFrameCount++;
                        
                        if(encodedVideoFrameCount == videoNbSample) onVideoEncodingComplete();
                        else if(waitingVideoReading) continueReading();

                        


                    },
                    error:(err)=>{
                        console.log("VideoEncoder error : ",err);
                    }

                });

                videoEncoder.configure(config);

                 
            }

            //--- setup audioEncoder ------

            let setupAudioEncoder = (config)=>{

                audioEncoder = new window["AudioEncoder"]({
                    output:(encodedChunk,config)=>{
                        if(config) console.log("AudioEncoder config ",config);
                    },
                    error:(err)=>{
                        console.log("AudioEncoder.error : ",err);
                    }
                })

                audioEncoder.configure(config);

            }

            



            //---- setup videoDecoder -----

            let getExtradata = ()=>{ 
                // generate the property "description" for the object used in VideoDecoder.configure
                // This function have been written by Thomas Guilbert from Google

                let avccBox = file.moov.traks[0].mdia.minf.stbl.stsd.entries[0].avcC;

                let i,size = 7;
                for (i=0;i<avccBox.SPS.length; i++) size+= 2 + avccBox.SPS[i].length;
                for (i=0;i<avccBox.PPS.length; i++) size+= 2 + avccBox.PPS[i].length;

                let id = 0;
                let data = new Uint8Array(size);

                let writeUint8 = (value)=>{
                    data.set([value],id);
                    id++;
                }
                let writeUint16 = (value)=>{
                    let arr = new Uint8Array(1);
                    arr[0] = value;
                    let buffer = new Uint8Array(arr.buffer);
                    data.set([buffer[1], buffer[0]], id);
                    id += 2;
                }
                let writeUint8Array = (value)=>{
                    data.set(value,id);
                    id += value.length;
                }

                writeUint8(avccBox.configurationVersion);
                writeUint8(avccBox.AVCProfileIndication);
                writeUint8(avccBox.profile_compatibility);
                writeUint8(avccBox.AVCLevelIndication);
                writeUint8(avccBox.lengthSizeMinusOne + (63<<2));
                writeUint8(avccBox.nb_SPS_nalus + (7<<5));

                for (i = 0; i < avccBox.SPS.length; i++) {
                    writeUint16(avccBox.SPS[i].length);
                    writeUint8Array(avccBox.SPS[i].nalu);
                }
            
                writeUint8(avccBox.nb_PPS_nalus);
                for (i = 0; i < avccBox.PPS.length; i++) {
                    writeUint16(avccBox.PPS[i].length);
                    writeUint8Array(avccBox.PPS[i].nalu);
                }

                if(id != size) throw "size mismatched !"
                return data;
            }



            let setupVideoDecoder = (config)=>{
                
               
                let timeout = null;

                processingVideo = true;
                waitingFrame = true;
                countSample = 0;
                nbSampleTotal = videoTrack.nb_samples;

                output.width = outputW;
                output.height = outputH;

                videoDecoder = new window["VideoDecoder"]({
                    output:(videoFrame)=>{
                       
                        createImageBitmap(videoFrame).then((img)=>{
                            
                            videoFrames.push(img);     
                            videoFrame.close();
                            freeMemory();

                            

                            //I use a timeout to give some time to videoDecoder to provide a bunch of frames
                            //The idea behind this is to maintain a small amount of active frames in memory. 
                            //
                            //=> videoDecoder output frame by frame but I can't compare the amount of
                            //frame decoded by VideoDecoder and the amount of frame sent to VideoDecoder
                            //because the first bunch of frames released by VideoDecoder is always smaller than
                            //the amount of frame sent.
                            //
                            //the amount of frames released at the beginning depends of the video file itself, it's not
                            //a fixed value.
                            //===> the difference will be released at the end of the decoding using VideoDecoder.flush
                            //
                            //Without setTimeout, the process start again once I got a single frame.
                            //=> I extract new frames once the array "videoFrames" is empty, so if i continue the process
                            //immediatly after I got a single frame, the array become empty almost immediatly and the extraction
                            //provide a lot of frame too early compared to the real progress of the video-reading

                            if(timeout) clearTimeout(timeout);
                            timeout = setTimeout(() => {
                                if(waitingFrame){
                                    waitingFrame = false;
                                    continueReading();
                                }
                            }, 15);
                            
                        })

                        if(!this.started){
                            this.started = true;
                            if(this.onReadyToPlay) this.onReadyToPlay();
                        }
                    },
                    error:(e)=>{
                        console.log("webcodec.VideoDecoder error : ",e);
                    }
                })

                videoDecoder.configure(config);
                file.setExtractionOptions(videoTrack.id,null,{nbSamples:nbSampleMax});
            }
            //----- setup AudioDecoder ---------

            let setupAudioDecoder = (config)=>{

                let timeout = null;
                countSample = 0;
                processingAudio = true;
                waitingFrame = true;
                nbSampleTotal = audioTrack.nb_samples;
                console.log("audio nb sample total = ",nbSampleTotal)

                audioDecoder = new window["AudioDecoder"]({
                    output:(audioFrame)=>{
                        //console.log("audioFrame = ",audioFrame);

                       

                        audioFrames.push(audioFrame);
                        


                        if(timeout) clearTimeout(timeout);
                        timeout = setTimeout(() => {
                            if(waitingFrame){
                                waitingFrame = false;
                                readNextFrame();
                            }
                        }, 15);
                    },
                    error:(err)=>{
                        console.log("WebCodec.AudioDecoder error : ",err);
                    }
                })

                audioDecoder.configure(config);

                file.setExtractionOptions(audioTrack.id,null,{nbSamples:nbSampleMax});


            }



            //----- setup mp4box and instanciate videoDecoder & videoEncoder ------

            file = MP4Box.createFile();
            
            file.onerror = (e)=>{
                console.log("file onerror ",e);
            }

            file.onError = (e)=>{
                console.warn("MP4Box file error => ",e);
            }
            file.onReady = (info)=>{
                muxStarted = true;
                videoTrack = info.videoTracks[0];
                audioTrack = info.audioTracks[0];
                console.log("audioTrack ",audioTrack)
                
                if(audioTrack){
                    audioSamplerate = audioTrack.audio.sample_rate;
                    audioChannelCount = audioTrack.audio.channel_count;
                    audioNbSample = audioTrack.nb_samples;
                }
                videoNbSample = videoTrack.nb_samples;
                videoDuration = (info.duration / info.timescale) * 1000;
                //|-> with some videos, videoTrack.movie_duration doesn't match with the real duration 

                videoFramerate = Math.ceil( 1000/(videoDuration / videoTrack.nb_samples) );
                videoFrameDurationInMicrosecond = oneSecondInMicrosecond / videoFramerate;
                //|-> using Math.ceil doesn't seem to be a good idea, I'll see that later... 

                videoW = videoTrack.track_width;
                videoH = videoTrack.track_height;

                outputW = videoW * encodingVideoScale;
                outputH = videoH * encodingVideoScale;


                console.log("videoFramerate ",videoFramerate)

                setupVideoEncoder({
                    codec : 'avc1.42001E',
                    width:outputW,
                    height:outputH,
                    hardwareAcceleration:"prefer-hardware", 
                    framerate:videoFramerate,
                    bitrate:15000000,
                    avc:{format:"avc"} 
                })

                setupVideoDecoder({
                    codec:videoTrack.codec,
                    codedWidth:videoW,
                    codedHeight:videoH,
                    description:getExtradata()
                })

                
               
                onVideoReadyToPlay(); 
                //=> at the bottom of the code , will call getNextSampleArray();
                //                               |===> will call file.start();
            }


            file.onSamples = (trackId, ref, samples)=>{
               
                //I process the dumux-step little by little in order to save memory 
                //so I stop file reading between 2 demux-process

                if(videoTrack.id == trackId){
                    stopped = true;
                    file.stop();


                    countSample += samples.length;

                    //console.log("onSample ",countSample+" VS "+nbSampleTotal)

                    for (const sample of samples) {
                        const type = sample.is_sync ? "key" : "delta";

                        const chunk = new window["EncodedVideoChunk"]({
                            type: type,
                            timestamp: sample.cts,
                            duration: sample.duration,
                            data: sample.data
                        });
                        
                        videoDecoder.decode(chunk);                   
                    }

                    if(countSample == nbSampleTotal){
                        videoDecoder.flush();
                    }

                    return
                }

                //-----------

                if(audioTrack.id == trackId){
                    //console.log("get audio sample")

                    stopped = true;
                    file.stop();
                    countSample += samples.length;

                    //console.log("onSample ",countSample+" VS "+nbSampleTotal)

                    for (const sample of samples) {
                        const type = sample.is_sync ? "key" : "delta";

                        const chunk = new window["EncodedAudioChunk"]({
                            type: type,
                            timestamp: sample.cts,
                            duration: sample.duration,
                            data: sample.data,
                            offset:sample.offset
                        });
                        
                        audioDecoder.decode(chunk);                   
                    }

                    if(countSample == nbSampleTotal){
                        audioDecoder.flush();
                    }
                }
                
                

            }


            let loadFile = (url)=>{
                
                fetch(url).then((response)=>{
                    //we fill our Mp4BoxFile with the data of our video file
                    
                    let offset = 0;
                    let buf;
                    let reader = response.body.getReader();

                    let push = ()=>{
                        return reader.read().then(({ done, value }) => {
                            
                            if(done == true) {
                                file.flush(); //-> will call file.onReady
                                return;
                            }

                            buf = value.buffer;
                            buf.fileStart = offset;
                            offset += buf.byteLength;
                            file.appendBuffer(buf);
                            push();
                        }).catch((e)=>{
                            console.log("reader error ",e)
                        })
                        
                    };
                    push();

                })
            }
           

            //--------------------------

            
            
            let onVideoReadyToPlay = ()=>{
                console.log("videoReadyToPlay")
                getNextSampleArray();
            }
            
            loadFile(url);


        </script>
    </body>
</html>
```

Thank you for your help and all your great work ! 

https://user-images.githubusercontent.com/272817/133670355-24dcd7ea-260c-414c-975b-b73b315016a6.mp4

**Comment by cconcolato** [comment]
I'll try to have a look but regarding audio, I agree. It's hard to find any example. My experiments in Google Canary indicates that AAC is not supported, only Opus:

```
AudioEncoder.isConfigSupported({codec: "mp4a.40.2", sampleRate: 48000, numberOfChannels: 1, bitrate: 96000})
Promise {<fulfilled>: {…}}[[Prototype]]: Promise[[PromiseState]]: "fulfilled"[[PromiseResult]]: 
Objectconfig: {bitrate: 96000, codec: 'mp4a.40.2', numberOfChannels: 1, sampleRate: 48000}
supported: false
[[Prototype]]: Object
```
```
AudioEncoder.isConfigSupported({codec: "opus", sampleRate: 48000, numberOfChannels: 1, bitrate: 96000})
Promise {<fulfilled>: {…}}[[Prototype]]: Promise[[PromiseState]]: "fulfilled"[[PromiseResult]]: 
Objectconfig: {bitrate: 96000, codec: 'opus', numberOfChannels: 1, sampleRate: 48000}
supported: true
[[Prototype]]: Object
```

**Comment by tlecoz** [comment]
Hello @cconcolato  !

Thank you for your message, I'm too exhausted tonight to work again but I'll try to complete the audioEncoder part tomorrow. 

In your opinion, if I do something like that in audioEncoder.output, will it work ? 

```

                       if(encodingAudioTrack == null){
                            audioEncodingTrackOptions.avcDecoderConfigRecord = config.decoderConfig.description;
                            encodingAudioTrack = outputFile.addTrack(audioEncodingTrackOptions);
                        }

                        let buffer = new ArrayBuffer(encodedChunk.byteLength);
                        encodedChunk.copyTo(buffer);

                        audioEncodingSampleOptions.dts = audioEncodingSampleOptions.cts = encodedChunk.timestamp;
                        audioEncodingSampleOptions.is_sync = encodedChunk.type == "key";

                        outputFile.addSample(encodingAudioTrack,buffer,audioEncodingSampleOptions);

```

it's the exact same code than the one used for videoEncoding

**Comment by tlecoz** [comment]
and just by curiosity, where did you find the valid codec string for AudioEncoder ? I searched a lot but found nothing... 

Thank you again !

**Comment by cconcolato** [comment]
The `avcDecoderConfigRecord` should not be used for audio. It's only for AVC videos. But for the rest, yes it should work in my opinion.
I found the `opus` example by chance!

**Comment by tlecoz** [comment]
Great ! I'll try that tomorrow and I'll come back here with the completed version of the code 

Do you have any idea concerning the "playhead problem" (when I read the mp4 file generated by mp4box inside a classic video player, the playhead doesn't move - but the movie plays correctly - )

**Comment by tlecoz** [comment]
@cconcolato  
Oh ! excuse me, i have another question  

In my code you can see these lines (in file.onReady )

```
videoDuration = (info.duration / info.timescale) * 1000;
//|-> with some videos, videoTrack.movie_duration doesn't match with the real duration  
```

Is it normal that videoTrack.movie_duration doesn't match all the time with the real duration of the video ? 
The calcul (info.duration / info.timescale) * 1000 works  but I was very surprised to see that videoTrack.movie_duration was not reliable. 

if you want i can send you a video that doesn't work

**Comment by tlecoz** [comment]
This one for example : the real duration is 5 seconds but inside Mp4Box with videoTrack.movie_duration I get 450 000 with a timescale set to 90 000 

so, with my calcul ( 450 000 / 90 000) * 1000 = 5000 

(please excuse me if its a noob question , I just was surprised because movie_duration was reliable with every video I used until I try to use this one)

https://user-images.githubusercontent.com/272817/133694126-a4b527a8-0b99-4ebd-b509-abe07436a5f3.mp4

**Comment by Djuffin** [comment]
> * There is no example at all on the internet using AudioEncoder, can you give me a valid codec please

@tlecoz  
Sorry about the absence of audio demos.
Here is the encoding / decoding demo for opus: https://wc-audio-gen.glitch.me/

**Comment by tlecoz** [comment]
@Djuffin 
Don't be sorry, thank you for the great work (and for the demo ! :) )

**Comment by rse** [comment]
Your recent changes from 0.4.8 to 0.4.9 related to this issue breaks MP4Box.js for decoding video streams (e.g. feeded via FFMpeg).

The reason is that `trak.first_dts` in https://github.com/gpac/mp4box.js/blob/4b775ec3fecee06581a83c04ca6bde580061b605/src/isofile-advanced-creation.js#L211 is undefined when the sample is not generated (as in this issue) but feeded via an existing stream as it looks. At least there seem to be samples created within MP4Box.js which are not feeded via `addSample()`. 

Additionally, the code in added line in https://github.com/gpac/mp4box.js/blob/4b775ec3fecee06581a83c04ca6bde580061b605/src/isofile-advanced-creation.js#L183 also seems to be wrong to me: `options.dts` should be `sample.dts` or you set `first_dts` to `undefined` instead of the fallback value of `0`.

**Comment by rse** [comment]
To get the latest 0.4.9 working again for decoding an existing externally feeded stream, as a quick fix, I had to apply the following patch:

```
-	traf.add("tfdt").set("baseMediaDecodeTime", (sample.dts - trak.first_dts));
+	traf.add("tfdt").set("baseMediaDecodeTime", (sample.dts - (trak.first_dts || 0)));
```

**Comment by cconcolato** [comment]
@rse Thanks for noticing and fixing the issue. I've pushed your patch.

**Comment by ringcrl** [comment]
```js
BoxParser.createSampleEntryCtor(BoxParser.SAMPLE_ENTRY_TYPE_AUDIO, 	"opus");
```

![image](https://user-images.githubusercontent.com/16740021/139609532-07c9ee77-0477-4055-959c-00dc09debaa4.png)

Finally I managed to compose a video with opus audio track, it works well on Chrome and ffplay, but it can't be played by QuickTime on mac, I hope webcodecs will support acc soon.

**Comment by tlecoz** [comment]
@ringcrl Great news !!!! 
Can you post a working example here or somewhere on the internet to let other people try it directly ? 
Maybe @chcunningham could add it on this page 
https://w3c.github.io/webcodecs/samples/

@chcunningham I will test it again today and I will post it on https://crbug.com/new if the problem still exists. Sorry for the lack of response I was over-busy

Have a nice day everyone

**Comment by ringcrl** [comment]
> @ringcrl Great news !!!! Can you post a working example here or somewhere on the internet to let other people try it directly ? Maybe @chcunningham could add it on this page https://w3c.github.io/webcodecs/samples/
> 
> @chcunningham I will test it again today and I will post it on https://crbug.com/new if the problem still exists. Sorry for the lack of response I was over-busy
> 
> Have a nice day everyone

@tlecoz  Of course, you can run this demo with a little modification to mp4box：https://github.com/ringcrl/webcodecs-playground/tree/master/public/demo-mp4box

**Comment by tlecoz** [comment]
@ringcrl  unfortunatly your demo doesn't work on my computer : the output video file works but without audio track 
(i copy paste your code , downloaded your file mp4box.all.js , and even the video file you use , but it doesn't work :'( )

**Comment by ringcrl** [comment]
> @ringcrl unfortunatly your demo doesn't work on my computer : the output video file works but without audio track (i copy paste your code , downloaded your file mp4box.all.js , and even the video file you use , but it doesn't work :'( )

Is it through ffprobe that you see no track content?

I can see the content of the track as shown in the picture above, but I can't play it through the player that comes with the system, but I can hear the sound after dragging and dropping the mp4 file to Chrome @tlecoz

**Comment by tlecoz** [comment]
"Is it through ffprobe that you see no track content?" 
No, I just tryed to read it using windows media player

"but I can hear the sound after dragging and dropping the mp4 file to Chrome"
you're right, it works as expected in chrome. Even the "playhead" move correctly in chrome ! 

it's so frustrating that it doesn't work as expected inside a classic media player... 

As I told you this morning by email, a solution could be to create one file for the video, another one for the audio and "mix" them in a single file using ffmpeg-wasm but this lib is not lightweight at all... (20Mo) 

ffmpeg-wasm supports opus format and output a mp4 readable everywhere. I tryed one minute ago and it works

**Comment by ringcrl** [comment]
> "Is it through ffprobe that you see no track content?" No, I just tryed to read it using windows media player
> 
> "but I can hear the sound after dragging and dropping the mp4 file to Chrome" you're right, it works as expected in chrome. Even the "playhead" move correctly in chrome !
> 
> it's so frustrating that it doesn't work as expected inside a classic media player...
> 
> As I told you this morning by email, a solution could be to create one file for the video, another one for the audio and "mix" them in a single file using ffmpeg-wasm but this lib is not lightweight at all... (20Mo)
> 
> ffmpeg-wasm supports opus format and output a mp4 readable everywhere. I tryed one minute ago and it works

After spending a lot of time carefully piecing together the data against the opus documentation and resolving all the reported errors, it was frustrating to find out that it couldn't be played in its own player, but inadvertently dragged it to Chrome and found that it could actually be played.

I also hope someone can tell me what I need to do to play the audio in the system player, there are no more errors to guide me to fix it, and it looks like I'm just a little bit away from success!

**Comment by chcunningham** [comment]
> it's so frustrating that it doesn't work as expected inside a classic media player...

It may just be that the platform media player down't natively support opus. For ex, for windows media player I see some google results that suggest opus is not supported.

**Comment by tlecoz** [comment]
"It may just be that the platform media player down't natively support opus. For ex, for windows media player I see some google results that suggest opus is not supported." 


Probably but it's really hard to understand for us that Webcodec use a codec that works almost nowhere. If it doesn't work in windows media player nor in quicktime, well, it's almost like if it doesn't work at all...

**Comment by chcunningham** [comment]
Fragmented support for media codecs is unfortunately a well known issue. Opus is broadly supported by browsers, but default installed OS players appear to be lagging. FWIW, AAC encoding remains on WebCodecs near-term roadmap, but it may not be available on all platforms.

**Comment by jozefchutka** [comment]
Hi @tlecoz ,

you mention you are able to mux your project with ffmpeg.

I am having hard times to mux raw vp8 or h264 bytes from VideoEncoder by ffmpeg.

When just concatenating  VideoEncoder output chunks into a buffer and saving to a file video.h264, `ffmpeg -i video.h264` complains with various errors.

I wonder how you managed to resolve it ffmpeg.

**Comment by tlecoz** [comment]
@jozefchutka  

my nodejs app is based on Puppeteer , the code is written using Typescript

The part of the code that uses VideoEncoder is on the "client side" (inside a chrome browser driven by puppeteer). 
the ffmpeg side runs into nodejs. 

I'm sorry I can't send you the whole running code because it's a part of a very complex professionnal project, but I can give you the key-steps  

you must define some functions to make a bridge between chrome-brower driven by Puppeteer and NodeJs. 
it looks like that : 
```

page.exposeFunction("startMp4Encoding",(w:number,h:number,fps:number,nbFrameTotal:number,outputFilePath:string,audioFilePath:string)=>{

	getEncodingLimits(Math.max(w , h))
	
	/*
	the function called "getEncodingLimits" is used to determine the distance (in amount of frames) ffmpeg can handle between the last frame received by ffmpeg and the last frame encoded by VideoEncoder. The ffmpeg process is asynchronous and if you send  too much frames too quickly it will become very very slow (it's the same for VideoEncoder by the way), so when VideoEncoder has sent too much frame, you must pause the VideoEncoder process and resume it once Ffmpeg is ready to handle next frames. 
	This limit depends on the resolution of the frame : ffmpeg can handle much more frames in 1920x1080 than in 4k , and it can handle much more frames in 1080x720 than in 1920x1080. 
	*/
	
	return new Promise((onEncodingStarted,onError)=>{
		ffmpegEncoder.startVideoEncoding(w,h,fps,nbFrameTotal,outputFilePath,audioFilePath,()=>{
			console.log("Ffmpeg completed -> file created !"
		})
	})
})

page.exposeFunction("encodeVideoChunk",(base64Chunk:string)=>{
	return new Promise((onChunkEncoded,onError)=>{
		ffmpegEncoder.encodeVideoChunk(base64Chunk);
		onChunkEncoded(true);
	})
})
page.exposeFunction("endMp4Encoding",()=>{
	return new Promise((onEncodingEnded,onError)=>{
		ffmpegEncoder.end();
		onEncodingEnded(true);
	})
})   
```

These functions will be callable from the object "window" in the browser side


----------------------

ffmpegEncoder is a class I wrote on the nodejs side to drive ffmpeg
The functions inside it looks like that 

```
var processManager = require("child_process")
var atob = require('atob');

public startVideoEncoding(w:number,h:number,fps:number,nbFrameTotal:number,outputFilePath:string,audioFilePath:string,onFfmpegClose:()=>void){
	
	this.nbVideoFrameEncoded = 0;
	this.nbVideoFrameTotal = nbFrameTotal;
	this.captureEnabled = true;
	this.videoEncoding = true;
	this.minSocketBufferLength = w * h * 5;

	let command = "";
	let useGPU = true;

	if(useGPU){
		command += "-y "; 
		command += "-vsync 1 -hwaccel cuda -hwaccel_output_format cuda ";
		command += "-r "+fps+" ";
		command += "-i - "
		if(audioFilePath != "null") command += "-i "+audioFilePath+" ";
		command += "-c:v h264_nvenc ";
		command += "-preset p6 ";
		command += "-tune hq ";
		command += "-b:v 5M ";
		command += "-bufsize 5M "
		command += "-maxrate 10M "
		command += "-pixel_format yuv444p -preset default ";
		command += outputFilePath
	}else{
		command += "-y ";
		command += "-c:v h264 ";
		command += "-r "+fps+" ";
		command += "-i - ";
		if(audioFilePath != "null") command += "-i "+audioFilePath+" ";
		command += outputFilePath;//"./OUTPUT.mp4";
	}
	
	/*
	the commande "-i - " means that I send the frames to ffmpeg one by one (when I receive the chunks from VideoEncoder)
	I don't use a temporary file 'video.h264' like you do 
	it's faster to do it this way (i used your approach before :) ) 
	*/

	this.launchFfmpeg(command,this.debug,onFfmpegClose);
}

public curFFmpegFrame:number=0;
private launchFfmpeg(command:string,debug:boolean=false,onClose:()=>void=null){
	this.ffmpeg = processManager.spawn("ffmpeg",command.split(" "));
	this.stdin = this.ffmpeg.stdin;
	
	this.ffmpeg.on("close",()=>{
	//    this.ffmpeg.kill();
		if(onClose) onClose();
	})

	this.ffmpeg.stderr.on("data",(data)=>{
		if ((data!=null) && (data.indexOf("frame=")!=-1)) {
			let ff:string=String(data).substring(data.indexOf("frame=")+6,data.indexOf("fps=")).trim();
			this.curFFmpegFrame=parseInt(ff); //<== get the amount of frame already processed by ffmpeg
		}
		if(debug) console.log(""+data);
	})
}


private base64ToArrayBuffer(base64:string):ArrayBuffer{
	let binary_string = atob(base64);//window.atob(base64);
	let len = binary_string.length;
	let bytes = new Uint8Array(len);
	for (let i = 0; i < len; i++) bytes[i] = binary_string.charCodeAt(i);
	return bytes.buffer;
}

public encodeVideoChunk(base64Chunk:string){
	
	let captureEnabled = true;

	if(!this.stdin) captureEnabled=false;
	else if (this.stdin.connecting) captureEnabled=false;
	
	if (this.stdin.writableLength > this.minSocketBufferLength) {
		if (this.stdin.writableCorked==0) this.stdin.cork();
	} else {
		//console.log("throttle encoding! writableLength:"+this.stdin.writableLength)
		if (this.stdin.writableCorked > 0) this.stdin.uncork();
		captureEnabled=false;
	}

	if(!captureEnabled){
		if(this.captureEnabled){
			this.captureEnabled = false;
			if(this.onVideoEncodingNeedPause) this.onVideoEncodingNeedPause();
		}
	}else{
		if(!this.captureEnabled){
			this.captureEnabled = true;
			if(this.onVideoEncodingCanResume) this.onVideoEncodingCanResume();
		}
	}

	this.nbVideoFrameEncoded++;
	this.stdin.write(new Uint8Array(this.base64ToArrayBuffer(base64Chunk)));
	this.captureEnabled=true;
	if ((this.readCompleted) && (this.nbVideoFrameEncoded >= this.nbVideoFrameTotal)) {
		console.log("all finished!")
		this.stdin.end();
		return;
	}
}

public end(){
	if(!this.readCompleted){
		this.corkable = false
		this.captureEnabled = true;
		this.readCompleted = true; 
	}
}
```


------------------ 

in the browser side, you have the VideoEncoder  
when you start the VideoEncoder, you also must start FfmpegEncoder like that

`window["startMp4Encoding"](w,h,fps,nbFrameTotal,outputFilePath,audioFilePath);`

Inside the VideoEncoder.output function , you receive some encodedChunks. 
Store the buffer contained Inside the chunk into an ArrayBuffer like that :
```
let buffer = new Int8Array(encodedChunk.byteLength);
encodedChunk.copyTo(buffer);
```

In order to transfer the data from chrome-browser to nodejs, we need to convert the buffer into a base64 string like that   
```
private arrayBufferToBase64(buffer:ArrayBuffer):string{
   let binary = '';
   let bytes = new Uint8Array( buffer );
   let len = bytes.byteLength;
   for (let i = 0; i < len; i++) binary += String.fromCharCode( bytes[ i ] );
   return window.btoa(binary);
} 

let base64 = arrayBufferToBase64(buffer);
window["encodeVideoChunk"](base64); 
```

when every frames has been sent to ffmpeg, you must tell ffmpeg to stop to wait for new frames like that 
`window["endMp4Encoding"]();`



-------

As I said, sorry but I can't give you the whole running code, you must fill the blanks by yourself but I gave you the main parts. 

You need to build the system that pause & resume VideoEncoder (and actually, you must do this part also on the code that send frames to VideoEncoder because if you sent too much frame too quickly VideoEncoder will also become very slow (it may easily happen if you encode a long file). 

Hope it helps :)

**Comment by jozefchutka** [comment]
Perfect, thanks

**Comment by 6l3m** [comment]
> Hi @tlecoz ,
> 
> you mention you are able to mux your project with ffmpeg.
> 
> I am having hard times to mux raw vp8 or h264 bytes from VideoEncoder by ffmpeg.
> 
> When just concatenating VideoEncoder output chunks into a buffer and saving to a file video.h264, `ffmpeg -i video.h264` complains with various errors.
> 
> I wonder how you managed to resolve it ffmpeg.

You need to give ffmpeg an output with extension declaring the container you wanna mux into, and I would suggest you skip the encoding part since it's already taken care of. The command would then be:
`ffmpeg -i video.h264 -c copy output.mp4`

**Comment by jozefchutka** [comment]
Thanks for the heads up @6l3m , also worth to share that h264 stream for ffmpeg works with the following VideoEncoder config:

```
const config = {codec: "avc1.42001E", avc : { format: "annexb" },
ffmpeg -i input.h264 -c copy output.mp4
```

vp8 frames can not be streamed/consumed as simple, however a simple [IVF container](https://wiki.multimedia.cx/index.php?title=IVF) would do the job.

```
ffmpeg -i input.ivf -c copy output.webm
```

**Comment by cconcolato** [comment]
Glad that this thread helped people, but I'm closing it as it does not seem related anymore to mp4box.js. Reopen or open a new issue if you have problems with mp4box.js.

**Comment by hughfenghen** [comment]
I created a sample project,  use mp4box.js + webcodec mux to mp4 file and play it with MSE, maybe it can help some people.

The data flow:   
canvas ImageBitmap -> mp4box.js + webcodecs (run in webworker) -> ReadableStream(mp4 ArrayBuffer, can be pushed to the server) -> mp4box.js -> Segment ArrayBuffer -> MSE

https://github.com/hughfenghen/wcm-recoder

**Comment by hughfenghen** [comment]
I've released an npm lib that records MediaStream, saves data to mp4 file, and supports video and audio tracks.

```ts
import { AVRecorder } from '@webav/av-recorder'

const recorder = new AVRecorder(mediaStream, {
  width: 1280,
  height: 720,
  // aac or opus (default)
  audioCodec: 'aac'
})
await recorder.start()

const fileHandle = await window.showSaveFilePicker({
  suggestedName: `av-recorder.mp4`
})

recorder.outputStream
  .pipeTo(await fileHandle.createWritable())

// await recorder.stop()
```

https://github.com/hughfenghen/WebAV/blob/main/packages/av-recorder/README.md

**Comment by tlecoz** [comment]
thanks !

**Comment by chipweinberger** [comment]
Can anyone help with this? https://github.com/gpac/mp4box.js/issues/375

Thanks in advance.

**Comment by hughfenghen** [comment]
> trackOptions.avcDecoderConfigRecord = config.decoderConfig.description

Haha, the avcDecoderConfigRecord is frequently overlooked.

**Comment by tlecoz** [comment]
By the way, I made a library very easy to use to handle video encoding (with and without audio) faster than requestAnimationFrame 

https://github.com/tlecoz/Mp4Maker

you can see a demo here : https://github.com/tlecoz/Mp4Maker
